src/_pytest/mark/structures.py

```diff
@@ -283,28 +284,37 @@ def test_function():
     on the ``test_function`` object. """
 
     _config = None
+    _markers = set()
 
     def __getattr__(self, name):
         if name[0] == "_":
             raise AttributeError("Marker name must NOT start with underscore")
+
         if self._config is not None:
-            self._check(name)
+            self._update_markers(name)
+            if name not in self._markers:
+                warnings.warn(
+                    "Unknown pytest.mark.%s - is this a typo?  You can register "
+                    "custom marks to avoid this warning - for details, see "
+                    "https://docs.pytest.org/en/latest/mark.html" % name,
+                    UnknownMarkWarning,
+                )
+                if self._config.option.strict:
+                    fail("{!r} not a registered marker".format(name), pytrace=False)
+
         return MarkDecorator(Mark(name, (), {}))
 
-    def _check(self, name):
-        try:
-            if name in self._markers:
-                return
-        except AttributeError:
-            pass
-        self._markers = values = set()
-        for line in self._config.getini("markers"):
-            marker = line.split(":", 1)[0]
-            marker = marker.rstrip()
-            x = marker.split("(", 1)[0]
-            values.add(x)
+    def _update_markers(self, name):
+        # We store a set of registered markers as a performance optimisation,
+        # but more could be added to `self._config` by other plugins at runtime.
+        # If we see an unknown marker, we therefore update the set and try again!
```

Comment:
This comment doesn't seem correct, we do not try to check markers again on this function, also not in `__getattr__` as well (see my comment there).